File: LinkedList/RotateList.go
// LeetCode 61 - Rotate List
// Time Complexity: O(n) | Space Complexity: O(1)
package main

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	length := 0
	for current := head; current != nil; current = current.Next {
		length++
	}

	k %= length
	if k == 0 {
		return head
	}

	fast := head
	slow := head

	for k > 0 {
		fast = fast.Next
		k--
	}

	for fast.Next != nil {
		fast = fast.Next
		slow = slow.Next
	}

	answer := slow.Next
	slow.Next = nil
	fast.Next = head
	return answer
}

File: LinkedList/RotateList.py
# LeetCode 61 - Rotate List
# Time Complexity: O(n) | Space Complexity: O(1)
from typing import Optional


class Solution:
    def rotateRight(self, head: Optional["ListNode"], k: int) -> Optional["ListNode"]:
        if head is None or head.next is None:
            return head

        length = 0
        current = head
        while current is not None:
            length += 1
            current = current.next

        k %= length
        if k == 0:
            return head

        fast = head
        slow = head

        while k > 0:
            fast = fast.next
            k -= 1

        while fast.next is not None:
            fast = fast.next
            slow = slow.next

        answer = slow.next
        slow.next = None
        fast.next = head
        return answer
